docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -283,41 +314,71 @@ When the `CONSTRAINTS` option doesn't contain `count`, `REPLICAS` must be specif
+ALTER TABLE t1 PLACEMENT CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=1, CONSTRAINTS="[+zone=sh]" ROLE=follower REPLICAS=1, CONSTRAINTS="[+zone=gz]" ROLE=follower REPLICAS=1;
+# Only rule #3 applies for these examples
+CREATE TABLE t1 (a int) PLACEMENT POLICY `mycompanypolicy`;
+ALTER TABLE t1 PLACEMENT POLICY `mycompanypolicy`;
+```
+
+1. An impossible policy such as violating raft constraints (2 leaders) should result in an error:
+```sql
+# error
+ALTER TABLE test
+	ALTER PLACEMENT POLICY CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=2;
+```
+2. A policy that has an incorrect number of instances for a raft group (i.e. 1 leader and 1 follower, or just 1 leader and no followers) should be a warning. This allows for some transitional topologies and developer environments.
+3. A policy that is impossible based on the current topology (zone=sh and replicas=3, but there is only 1 store in sh) should be a warning. This allows for some transitional topologies.
+
+#### Ambiguous and edge cases
+
+The following two policies are not identical:
 
 ```sql
-ALTER {DATABASE | SCHEMA} schema_name
-	{ADD | ALTER} PLACEMENT POLICY ROLE=role CONSTRAINTS=constraints REPLICAS=replicas,
-	...
-	
-ALTER {DATABASE | SCHEMA} schema_name
-	DROP PLACEMENT POLICY ROLE=role,
-	...
+CREATE PLACEMENT POLICY p1 CONSTRAINTS="[+zone=sh]" ROLE=follower REPLICAS=1, CONSTRAINTS="[+zone=gz]" ROLE=follower REPLICAS=1;
+CREATE PLACEMENT POLICY p2 CONSTRAINTS="[+zone=sh,+zone=gz]" ROLE=follower REPLICAS=2;
 ```
 
-This statement defines placement rules for one database, including all tables in it.
+This is because "[+zone=sh,+zone=gz]" is interpreted as either sh or gz. It is possible that both replicas could end up in the same zone. This is useful in the case that you want to ensure that `REPLICAS=2` exists in any of a list of zones:
 
-Creating or dropping a table also affects the placement rules. If a placement rule is defined on a database, all tables in this database will automatically apply that rule, including the existing tables and the tables created later.
+```sql
+CREATE PLACEMENT POLICY p2 CONSTRAINTS="[+zone=sh,+zone=gz,+zone=bz]" ROLE=follower REPLICAS=2;
+```
 
-Once the placement rules on a database are changed, the tables should also update their placement rules. Users can overwrite the rules by defining placement rules on the tables. See the section "Rule inheritance" for details.
+However, using "short hand" JSON syntax, the following 2 policies are identical:
 
-Since key range is not successive in one database, each table in the database corresponds to at least one placement rule, so there may be many placement rules.
+```sql
+ALTER TABLE test
+	PLACEMENT POLICY CONSTRAINTS="{+zone=sh:2,+zone=bj:2}" ROLE=voter REPLICAS=4;
+
+ALTER TABLE test
+	PLACEMENT POLICY CONSTRAINTS="[+zone=sh]" ROLE=voter REPLICAS=2,
+	PLACEMENT POLICY CONSTRAINTS="[+zone=bj]" ROLE=voter REPLICAS=2;
```

Comment:
So the the expression `+label1=value1,+label1=value2,+label2=value3,+label2=value4,-label3=value5,-label3=value6` means `(label1 in (value1, value2)) and (label2 in (value3, value4)) and (label3 notin (value5, value6))`. We apply `OR` for labels with same name and `AND` for labels with different name.  Is that right?

Reply:
No. For you can not schedule one follower into both `sh` and `bj`, it is banned. We can workaround later. But I agree that the semantics should be clarified.

Reply:
But how to interpret this sentence above: "This is because "[+zone=sh,+zone=gz]" is interpreted as either sh or gz..."

Reply:
In one `CONSTRAINTS="..."`, it is banned. But you can specify another different `CONSTRAINTS`, that becomes `OR`.

Reply:
Anyway, I think it's ambiguous and hard to learn.

Reply:
This is now outdated because the proposal only has one set of constraints, and defines `constraints` + `follower_constraints` as AND.